File: untest/156.binary-tree-upside-down.java
public TreeNode upsideDownBinaryTree(TreeNode root) {
    if(root==null) return null;
    return doUpsideDown(root);
}
 
private TreeNode doUpsideDown(TreeNode root) {
    if(root.left==null) {
        return root;
    }
    TreeNode newRoot = doUpsideDown(root.left);
    root.left.right=root;
    root.left.left=root.right;
    root.left=root.right=null;
    return newRoot;
}

File: untest/156.binary-tree-upside-down.cc
BinaryTree* BinaryBottomUp(BinaryTree* root)
{
    doBottomUp(root);
    return root;
}
 
void doBottomUp(BinaryTree*& root)
{
    if(!root || !root->left) return;
    BinaryTree* ret=root->left;
    doBottomUp(ret);
    root->left->left=root->left;
    root->left->right=root->right;
    root->left=root->right=NULL;
    root=ret;
}
